File: archive/r/ruby/longest-palindromic-substring.rb
def expand_from_center(s, left, right)
  while left >= 0 && right < s.length && s[left] == s[right]
    left -= 1
    right += 1
  end
  s[(left + 1)...right]
end

def longest_palindromic_substring(s)
  abort "Usage: please provide a string that contains at least one palindrome" if s.to_s.empty?

  longest = ""

  s.length.times do |i|
    odd  = expand_from_center(s, i, i)
    even = expand_from_center(s, i, i + 1)

    longest = odd  if odd.length  > longest.length
    longest = even if even.length > longest.length
  end

  abort "Usage: please provide a string that contains at least one palindrome" if longest.length <= 1

  puts longest
end

input = ARGV.first
longest_palindromic_substring(input)

File: archive/s/swift/longest-palindromic-substring.swift
import Foundation

let usage = """
    Usage: please provide a string that contains at least one palindrome
    """

extension Collection {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

func findLongestPalindrome(_ s: String) -> String {
    let originalChars = Array(s)
    guard !originalChars.isEmpty else { return "" }

    let processed: [Character] = ["#"] + originalChars.flatMap { [$0, "#"] }

    let n = processed.count
    var radii = Array(repeating: 0, count: n)
    var center = 0
    var rightBoundary = 0

    var maxRadius = 0
    var centerOfMax = 0

    for i in 0..<n {
        if i < rightBoundary {
            let mirror = 2 * center - i
            radii[i] = min(rightBoundary - i, radii[mirror])
        }

        var left = i - (1 + radii[i])
        var right = i + (1 + radii[i])

        while left >= 0 && right < n, processed[left] == processed[right] {
            radii[i] += 1
            left -= 1
            right += 1
        }

        if i + radii[i] > rightBoundary {
            center = i
            rightBoundary = i + radii[i]
        }

        if radii[i] > maxRadius {
            maxRadius = radii[i]
            centerOfMax = i
        }
    }

    guard maxRadius > 1 else { return "" }

    let start = (centerOfMax - maxRadius) / 2
    let resultRange = start..<(start + maxRadius)

    return String(originalChars[resultRange])
}

guard let input = CommandLine.arguments[safe: 1], !input.isEmpty else {
    print(usage)
    exit(1)
}

let result = findLongestPalindrome(input)

if result.isEmpty {
    print(usage)
    exit(1)
} else {
    print(result)
}
